From fbbaa2f9fce05121e7eeac0c590067ea741af623 Mon Sep 17 00:00:00 2001
From: Andrea Alberti <a.alberti82@gmail.com>
Date: Fri, 4 Sep 2026 23:27:06 +0200
Subject: [PATCH] markdown-ts-mode: Recognize pointy-bracket link destinations
 (bug#81797)

CommonMark specifications (spec 0.31.2, section 6.3 "Links") allow link
destinations to be wrapped by pointy brackets `<...>'; wrapping
destinations by pointy brackets is needed when the destination includes
spaces.  This patch fixes markdown-ts-mode by recognizing destinations
wrapped by pointy brackets.  The fix simply consists of trimming away
the pointy brackets.

* lisp/textmodes/markdown-ts-mode.el
(markdown-ts--unbracket-destination): New function.
(markdown-ts--make-link-button): Unbracket URL used to create a
clickable link.
(markdown-ts--fontify-image): Unbracket the image destination too.
---
 lisp/textmodes/markdown-ts-mode.el | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/lisp/textmodes/markdown-ts-mode.el b/lisp/textmodes/markdown-ts-mode.el
index 0ab60f79dc3..24e9ad2d32c 100644
--- a/lisp/textmodes/markdown-ts-mode.el
+++ b/lisp/textmodes/markdown-ts-mode.el
@@ -886,6 +886,18 @@ markdown-ts--fontify-atx-delimiter
 
 (defvar url-mail-command) ; url/url-vars.el
 
+(defun markdown-ts--unbracket-destination (url)
+  "Return URL after stripping the optional pointy-bracket wrapper `<...>'.
+Pointy brackets are in general optional, but must be used for
+destinations containing spaces; see CommonMark 0.31.2, section 6.3
+\"Links\"."
+  ;; Strip the pointy brackets only when both delimiters are present,
+  ;; while leaving a destination such as `foo>' unchanged in a legitimate
+  ;; link like `[a](foo>)'.
+  (if (and (string-prefix-p "<" url) (string-suffix-p ">" url))
+      (substring url 1 -1)
+    url))
+
 (defun markdown-ts--make-link-button (beg end url)
   "Make the region from BEG to END a clickable button for URL.
 For mailto: URIs, use `url-mail-command'.  For other schemes
@@ -900,7 +912,8 @@ markdown-ts--make-link-button
 heading face."
   ;; NOTE: URI scheme and host name are case-insensitive per RFC 3986
   ;; and RFC 7230.
-  (let ((case-fold-search nil))
+  (let ((url (markdown-ts--unbracket-destination url))
+        (case-fold-search nil))
     (make-text-button beg end
                       'action (lambda (_button)
                                 (cond
@@ -1715,7 +1728,8 @@ markdown-ts--fontify-image
                ;; with the folded display.
                (not (markdown-ts--outline-invisible-p node-start)))
       (let* ((dest (treesit-search-subtree node "\\`link_destination\\'"))
-             (url (and dest (treesit-node-text dest t)))
+             (url (and dest (markdown-ts--unbracket-destination
+                             (treesit-node-text dest t))))
              (remotep (and url (string-match-p "\\`https?://" url)))
              (displayable
               (when url
-- 
2.55.0

